src/test/java/org/reaktivity/specification/socks/internal/SocksFunctionsTest.java

```diff
@@ -44,43 +48,136 @@ public void setUp() throws Exception
+        DirectBuffer buffer = new UnsafeBuffer(bytes);
+        SocksRouteExFW routeEx = new SocksRouteExFW().wrap(buffer, 0, buffer.capacity());
+
+        assertEquals("example.com", routeEx.address().domainName().asString());
+        assertEquals(8080, routeEx.port());
+    }
+
+    @Test
+    public void shouldBuildBeginExWithDomainName() throws Exception
+    {
+        byte[] bytes = SocksFunctions.beginEx()
+                                     .typeId(0)
+                                     .domain("example.com")
+                                     .port(8080)
+                                     .build();
+        DirectBuffer buffer = new UnsafeBuffer(bytes);
+        SocksBeginExFW beginEx = new SocksBeginExFW().wrap(buffer, 0, buffer.capacity());
+
+        assertEquals("example.com", beginEx.address().domainName().asString());
+        assertEquals(8080, beginEx.port());
+    }
+
+    @Test
+    public void shouldLoadFunctionsWithIpv4Address() throws Exception
+    {
+        String expressionText = "${socks:routeEx()" +
+                                       ".address(\"127.0.0.1\")" +
+                                       ".port(8080)" +
+                                       ".build()}";
+        ValueExpression expression = factory.createValueExpression(ctx, expressionText, String.class);
+        String routeExBytes = (String) expression.getValue(ctx);
+
+        assertNotNull(routeExBytes);
+    }
+
+    @Test
+    public void shouldBuildRouteExWithIpv4Address() throws Exception
+    {
+        byte[] bytes = SocksFunctions.routeEx()
+                                     .address("127.0.0.1")
+                                     .port(8080)
+                                     .build();
+        DirectBuffer buffer = new UnsafeBuffer(bytes);
+        SocksRouteExFW routeEx = new SocksRouteExFW().wrap(buffer, 0, buffer.capacity());
+
+        assertEquals(KIND_IPV4_ADDRESS, routeEx.address().kind());
+        assertArrayEquals(new byte[] { 127, 0, 0, 1 },
+            copyOfRange(routeEx.buffer().byteArray(), routeEx.address().offset() + 1, routeEx.address().limit()));
+        assertEquals(8080, routeEx.port());
+    }
+
+    @Test
+    public void shouldBuildBeginExWithIpv4Address() throws Exception
+    {
+        byte[] bytes = SocksFunctions.beginEx()
+                                     .typeId(0)
+                                     .address("127.0.0.1")
+                                     .port(8080)
+                                     .build();
+        DirectBuffer buffer = new UnsafeBuffer(bytes);
+        SocksBeginExFW beginEx = new SocksBeginExFW().wrap(buffer, 0, buffer.capacity());
+
+        assertEquals(KIND_IPV4_ADDRESS, beginEx.address().kind());
+        assertArrayEquals(new byte[] { 127, 0, 0, 1 },
+            copyOfRange(beginEx.buffer().byteArray(), beginEx.address().offset() + 1, beginEx.address().limit()));
+        assertEquals(8080, beginEx.port());
```

Comment:
Hoist `beginEx.address()` into separate variable to avoid repetitive calls to `beginEx.address()`.

After asserting `address.kind()` is `IPv4`, `assertArrayEquals(..., copyOfRange(address.ipv4Address().buffer().byteArray(), address.ipv4Address().offset(), address.ipv4Address().limit()))` to eliminate the magic `+1` for `offset` calculation.

Please apply same feedback to other similar tests.

Reply:
done